tests/unit/api/auth.py

```diff
@@ -0,0 +1,553 @@
+    (api.node_delete, ['no_nic_node']),
+    (api.node_register_nic, ['free_node_0', 'extra-nic', 'de:ad:be:ef:20:16']),
+    (api.node_delete_nic, ['free_node_0', 'boot-nic']),
+    (api.project_create, ['anvil-nextgen']),
+    (api.list_projects, []),
+
+    # node_power_*, on free nodes only. Nodes assigned to a project are
+    # tested in project_calls, below.
+    (api.node_power_cycle, ['free_node_0']),
+    (api.node_power_off, ['free_node_0']),
+
+    (api.project_delete, ['empty-project']),
+
+    (api.switch_register, ['new-switch', MockSwitch.api_name]),
+    (api.switch_delete, ['empty-switch']),
+    (api.switch_register_port, ['stock_switch_0', 'new_port']),
+    (api.switch_delete_port, ['stock_switch_0', 'free_port_0']),
+    (api.port_connect_nic, ['stock_switch_0', 'free_port_0',
+                            'free_node_0', 'boot-nic']),
+    (api.port_detach_nic, ['stock_switch_0', 'free_node_0_port']),
+]
+
+
+# Similarly, there are a large number of calls that require access to a
+# particular project. This is a list of (function, args) pairs that should
+# succeed as project 'runway', and fail as project 'manhattan'.
+project_calls = [
+    # node_power_*, on allocated nodes only. Free nodes are testsed in
+    # admin_calls, above.
+    (api.node_power_cycle, ['runway_node_0']),
+    (api.node_power_off, ['runway_node_0']),
+
+    (api.project_connect_node, ['runway', 'free_node_0']),
+    (api.project_detach_node, ['runway', 'runway_node_0']),
+
+    (api.headnode_create, ['new-headnode', 'runway', 'base-headnode']),
+    (api.headnode_delete, ['runway_headnode_off']),
+    (api.headnode_start, ['runway_headnode_off']),
+    (api.headnode_stop, ['runway_headnode_on']),
+    (api.headnode_create_hnic, ['runway_headnode_off', 'extra-hnic']),
+    (api.headnode_delete_hnic, ['runway_headnode_off', 'pxe']),
+
+    (api.headnode_connect_network, ['runway_headnode_off', 'pxe', 'stock_int_pub']),
+    (api.headnode_connect_network, ['runway_headnode_off', 'pxe', 'runway_pxe']),
+    (api.headnode_detach_network, ['runway_headnode_off', 'public']),
+
+    (api.list_project_headnodes, ['runway']),
+    (api.show_headnode, ['runway_headnode_on']),
+]
+
+
+@pytest.mark.parametrize('fn,args', admin_calls)
+def test_admin_succeed(fn, args):
+    auth_call_test(fn, None,
+                   True, None,
+                   args)
+
+
+@pytest.mark.parametrize('fn,args', admin_calls)
+def test_admin_fail(fn, args):
+    auth_call_test(fn, AuthorizationError,
+                   False, 'runway',
+                   args)
+
+
+@pytest.mark.parametrize('fn,args', project_calls)
+def test_runway_succeed(fn, args):
+    auth_call_test(fn, None,
+                   False, 'runway',
+                   args)
+
+
+@pytest.mark.parametrize('fn,args', project_calls)
+def test_manhattan_fail(fn, args):
+    auth_call_test(fn, AuthorizationError,
+                   False, 'manhattan',
+                   args)
+
```

Comment:
Yeah, I did the other stuff first. I'll take a second look at it to see if I can't improve it. I think I at least have an idea of how to use a dict.